src/Metadata/Resource/SubResourceMetadata.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Metadata\Resource;
+
+final class SubResourceMetadata
```

Comment:
Hmm this is a good idea. I didn't wanted to introduce a new Metadata class when I worked on subresources but this will give more power to subresource declarations (ping @teohhanhui  ^^).

WDYT about going a bit further, so that `PropertyMetadataFactory` will create the `SubResourceMetadata` when it's declared as such? This is what the outcome may be in the code below:

```php
$propertyMetadata = $this->propertyMetadataFactory->create($parentResourceClass, $property);
if ($propertyMetadata->hasSubresource()) {
        $subResourceMetadata = $propertyMetadata->getSubresource(); // this is an instance of SubResourceMetadata
}
```

With something like this we can avoid repeating the following code (for example also in `ApiLoader`): 

```
$isCollection = $propertyMetadata->getType()->isCollection();
$resourceClass = $isCollection ? $propertyMetadata->getType()->getCollectionValueType()->getClassName() : $propertyMetadata->getType()->getClassName();
```

ping @api-platform/core-team 

IMO we need to agree on this change first because it may involve many changes ^^. Anyway, if we introduce this new metadata class, it'd be a shame not to use it everywhere we can :).